File: 143_reorder_list.ts
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function printLL(head: ListNode | null): void {
    let curr = head;
    const arr: number[] = [];
    while (curr) {
        arr.push(curr.val);
        curr = curr.next;
    }
    console.log(arr);
}

function reverseLL(head: ListNode | null) {
    let prev: ListNode | null = null;
    let temp = head;

    while (head) {
        temp = head.next;
        head.next = prev;
        prev = head;
        head = temp;
    }

    return prev;
}

function reorderList(head: ListNode | null): void {
    if (!head) return;

    let slow: ListNode | null = head;
    let fast = head.next;

    while (fast !== null && fast.next !== null) {
        fast = fast!.next;
        fast = fast!.next;
        slow = slow!.next;
    }

    let secondHalf = slow!.next;
    slow!.next = null;

    secondHalf = reverseLL(secondHalf);

    let curr: ListNode | null = head;
    while (curr && secondHalf) {
        let temp1 = curr.next;
        let temp2 = secondHalf!.next;
        curr.next = secondHalf;
        secondHalf!.next = temp1;
        secondHalf = temp2;
        curr = temp1;
    }
}

File: 8thJune2024/143_reorder_list.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *reverseLL(ListNode *head) {
    ListNode *prev = nullptr, *temp, *curr;
    curr = head;
    while (curr != nullptr) {
      temp = curr->next;
      curr->next = prev;
      prev = curr;
      curr = temp;
    }
    return prev;
  }

  void printLL(ListNode *head) {
    ListNode *curr = head;
    while (curr != nullptr) {
      cout << curr->val << " ";
      curr = curr->next;
    }
    cout << endl;
  }

  void reorderList(ListNode *head) {
    ListNode *slow = head, *fast = head->next;
    while (fast != nullptr && fast->next != nullptr) {
      slow = slow->next;
      fast = fast->next;
      fast = fast->next;
    }
    ListNode *list2 = reverseLL(slow->next);
    slow->next = nullptr;
    ListNode *list1 = head;
    ListNode *temp1, *temp2;
    while (list1 != nullptr && list2 != nullptr) {
      temp1 = list1->next;
      temp2 = list2->next;
      list1->next = list2;
      list2->next = temp1;
      list1 = temp1;
      list2 = temp2;
    }
  }
};
